crates/lib/src/http.rs

```diff
@@ -0,0 +1,447 @@
+
+        let address = match host {
+            Host::Ipv4(ip) => SocketAddr::V4(SocketAddrV4::new(ip, port)),
+            Host::Ipv6(ip) => SocketAddr::V6(SocketAddrV6::new(ip, port, 0, 0)),
+            Host::Domain(domain) => {
+                format!("{}:{}", domain, port)
+                    .to_socket_addrs()
+                    .map_err(|e| HostLookup.context(e))?
+                    .next()
+                    .ok_or_else(|| HostLookup.context("No socket addresses found for host"))?
+            }
+        };
+
+        let http_settings = HttpSettings {
+            hostname: url.host_str(),
+            protocol: match url.scheme() {
+                #[cfg(feature = "tls")]
+                "https" => HttpProtocol::Https,
+                _ => HttpProtocol::Http,
+            },
+            headers: headers.unwrap_or_default(),
+        };
+
+        HttpClient::new(&address, timeout_settings, http_settings)
+    }
+
+    /// Send a HTTP GET request and return the response data as a buffer.
+    pub fn get(&mut self, path: &str) -> GDResult<Vec<u8>> { self.request("GET", path) }
+
+    /// Send a HTTP GET request and parse the JSON resonse.
+    pub fn get_json<T: DeserializeOwned>(&mut self, path: &str) -> GDResult<T> { self.request_json("GET", path) }
+
+    /// Send a HTTP Post request with JSON data and parse a JSON response.
+    pub fn post_json<T: DeserializeOwned, S: Serialize>(&mut self, path: &str, data: S) -> GDResult<T> {
+        self.request_with_json_data("POST", path, data)
+    }
+
+    // NOTE: More methods can be added here as required using the request_json or
+    // request_with_json methods
+
+    #[inline]
+    fn request(&mut self, method: &str, path: &str) -> GDResult<Vec<u8>> {
+        // Append the path to the pre-parsed URL and create a request object.
+        self.address.set_path(path);
+        let mut request = self.client.request_url(method, &self.address);
+
+        // Set the request headers.
+        for (key, value) in self.headers.iter() {
+            request = request.set(key, value);
+        }
+
+        // Send the request.
+        let http_response = request.call().map_err(|e| PacketSend.context(e))?;
+
+        let length = if let Some(length) = http_response.header("Content-Length") {
+            length
+                .parse::<usize>()
+                .map_err(|e| ProtocolFormat.context(e))?
+                .min(MAX_RESPONSE_LENGTH)
+        } else {
+            5012 // Sensible default allocation
+        };
+
+        let mut buffer: Vec<u8> = Vec::with_capacity(length);
+
+        let _ = http_response
+            .into_reader()
+            .take(MAX_RESPONSE_LENGTH as u64)
+            .read_to_end(&mut buffer)
+            .map_err(|e| PacketReceive.context(e))?;
+
+        Ok(buffer)
+    }
+
+    /// Send a HTTP request without any data and parse the JSON response.
+    #[inline]
```

Comment:
Hey, not a complain, but I got a curious about this, why have you inlined these functions?

Reply:
To hopefully reduce call cost of the request functions as they are all essentially wrappers of these (at the cost of binary size). It's a bit of a pre-mature optimization I have done no tests to confirm that this is better (the compiler might've inlined these anyway).